Mathlib/GroupTheory/MonoidLocalization.lean

```diff
@@ -1251,6 +1251,46 @@ theorem map_map {A : Type*} [CommMonoid A] {U : Submonoid A} {R} [CommMonoid R]
 #align submonoid.localization_map.map_map Submonoid.LocalizationMap.map_map
 #align add_submonoid.localization_map.map_map AddSubmonoid.LocalizationMap.map_map
 
+
+/-- Given an injective `CommMonoid` homomorphism `g : M →* P`, and a Submonoid `S ⊆ M`,
+the induced Monoid homomorphism from the Localization of `M` at `S` to the
+Localization of `P` at `g S`, is injective.
+-/
+@[to_additive "Given an injective `AddCommMonoid` homomorphism `g : M →+ P`, and a
+Submonoid `S ⊆ M`, the induced AddMonoid homomorphism from the Localization of `M` at `S`
+to the Localization of `P` at `g S`, is injective. "]
+theorem map_injective_of_injective (hg : Injective g)
+    (k : LocalizationMap (Submonoid.map g S) Q) :
+    Injective (map f (apply_coe_mem_map g S) k : N →* Q) := by
+  rw [Injective]
+  intro z w hizw
+  let i := (map f (apply_coe_mem_map g S) k : N→* Q)
+  have ifkg (a:M): i (f.toMap a) = k.toMap (g a) := map_eq f (apply_coe_mem_map g S) a
+  obtain ⟨x, hx⟩ := surj f z
+  obtain ⟨y, hy⟩ := surj f w
```

Comment:
I think when working with semirings (where you can't subtract and compare with zero) it's convenient to introduce a `surj₂` lemma that gives you a common denominator for both z and w. The same for `IsLocalization` and `IsLocalizedModule`. Would you like to do that?

Reply:
Do you mean something like 
```
theorem surj₂ (f : LocalizationMap S N) (z w : N) : ∃ x : M × M × S, 
    (z * f.toMap x.2.2 = f.toMap x.1) ∧  (w * f.toMap x.2.2 = f.toMap x.2.1) := by 
```

Reply:
Yes! Though I'm not a big fan of using existential over a product, and I think `∃ z' w' : M, ∃ x : S, …` is probably easier to use.

Reply:
I added that lemma, but I am not happy with the long proof.